components/server/ee/src/billing/billing-mode.ts

```diff
@@ -0,0 +1,121 @@
+/**
+ * Copyright (c) 2022 Gitpod GmbH. All rights reserved.
+ * Licensed under the GNU Affero General Public License (AGPL).
+ * See License-AGPL.txt in the project root for license information.
+ */
+
+import { inject, injectable } from "inversify";
+
+import { Team, User } from "@gitpod/gitpod-protocol";
+import { getExperimentsClientForBackend } from "@gitpod/gitpod-protocol/lib/experiments/configcat-server";
+import { SubscriptionService } from "@gitpod/gitpod-payment-endpoint/lib/accounting";
+import { Subscription } from "@gitpod/gitpod-protocol/lib/accounting-protocol";
+import { Config } from "../../../src/config";
+import { StripeService } from "../user/stripe-service";
+import { BillingMode } from "@gitpod/gitpod-protocol/lib/billing-modes";
+import { TeamSubscription2DB } from "@gitpod/gitpod-db/lib";
+
+@injectable()
+export class BillingModesImpl {
+    @inject(Config) protected readonly config: Config;
+    @inject(SubscriptionService) protected readonly subscriptionSvc: SubscriptionService;
+    @inject(StripeService) protected readonly stripeSvc: StripeService;
+    @inject(TeamSubscription2DB) protected readonly teamSubscription2Db: TeamSubscription2DB;
+
+    async getBillingModeForUser(user: User): Promise<BillingMode> {
+        if (!this.config.enablePayment) {
+            // Payment is not enabled. E.g. Self-Hosted.
+            return { mode: "none" };
+        }
+        const now = new Date();
+
+        // Is Usage Based Billing enabled for this user or not?
+        const isUsageBasedBillingEnabled = await getExperimentsClientForBackend().getValueAsync(
+            "isUsageBasedBillingEnabled",
+            false,
+            {
+                user,
+            },
+        );
+
+        // Any Chargebee subscriptions? This includes those derived from a Team Seat!
+        const cbSubscriptions = await this.subscriptionSvc.getActivePaidSubscription(user.id, now);
+        if (cbSubscriptions.length > 0) {
+            if (
+                isUsageBasedBillingEnabled &&
+                cbSubscriptions.every((s) => Subscription.isCancelled(s, now.toISOString()))
+            ) {
+                // The user has one or more paid subscriptions, but all of them have already been cancelled
+                return { mode: "chargebee", tier: "paid_cancelled_and_ubb" };
+            }
+
+            // The user has at least one paid subscription
+            return { mode: "chargebee", tier: "paid", planId: cbSubscriptions[0].planId! };
+        }
+
+        // UBB enabled?
+        if (!isUsageBasedBillingEnabled) {
+            // No Chargebee subscription, and UBB is not enabled: CB free tier?
```

Comment:
Why the question mark? 🤔

Reply:
I think it make sense for the two of us to discuss the decision tree/scope of the decision made here. :+1: